pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/WindowAggregateOperator.java

```diff
@@ -0,0 +1,431 @@
+    return ImmutableList.of(_inputOperator);
+  }
+
+  @Nullable
+  @Override
+  public String toExplainString() {
+    return EXPLAIN_NAME;
+  }
+
+  @Override
+  protected TransferableBlock getNextBlock() {
+    try {
+      if (!_readyToConstruct && !consumeInputBlocks()) {
+        return TransferableBlockUtils.getNoOpTransferableBlock();
+      }
+
+      if (_upstreamErrorBlock != null) {
+        return _upstreamErrorBlock;
+      }
+
+      if (!_hasReturnedWindowAggregateBlock) {
+        return produceWindowAggregateBlock();
+      } else {
+        // TODO: Move to close call.
+        return TransferableBlockUtils.getEndOfStreamTransferableBlock();
+      }
+    } catch (Exception e) {
+      return TransferableBlockUtils.getErrorTransferableBlock(e);
+    }
+  }
+
+  private boolean isPartitionByOnlyQuery(List<RexExpression> groupSet, List<RexExpression> orderSet,
+      List<RelFieldCollation.Direction> orderSetDirection,
+      List<RelFieldCollation.NullDirection> orderSetNullDirection) {
+    if (orderSet == null || orderSet.isEmpty()) {
+      return true;
+    }
+
+    if (groupSet == null || groupSet.isEmpty() || (groupSet.size() != orderSet.size())) {
+      return false;
+    }
+
+    Set<Integer> partitionByInputRefIndexes = new HashSet<>();
+    Set<Integer> orderByInputRefIndexes = new HashSet<>();
+    for (int i = 0; i < groupSet.size(); i++) {
+      partitionByInputRefIndexes.add(((RexExpression.InputRef) groupSet.get(i)).getIndex());
+      orderByInputRefIndexes.add(((RexExpression.InputRef) orderSet.get(i)).getIndex());
+    }
+
+    boolean isPartitionByOnly = partitionByInputRefIndexes.equals(orderByInputRefIndexes);
+    if (isPartitionByOnly) {
+      // Check the direction and null direction to ensure default ordering on the order by keys, which are:
+      // Direction: ASC
+      // Null Direction: LAST
+      for (int i = 0; i < orderSet.size(); i++) {
+        if (orderSetDirection.get(i) == RelFieldCollation.Direction.DESCENDING
+            || orderSetNullDirection.get(i) == RelFieldCollation.NullDirection.FIRST) {
+          isPartitionByOnly = false;
+          break;
+        }
+      }
+    }
+    return isPartitionByOnly;
+  }
+
+  private TransferableBlock produceWindowAggregateBlock() {
+    List<Object[]> rows = new ArrayList<>(_numRows);
+    for (Map.Entry<Key, List<Object[]>> e : _partitionRows.entrySet()) {
+      Key partitionKey = e.getKey();
+      List<Object[]> rowList = e.getValue();
+      for (Object[] existingRow : rowList) {
+        Object[] row = new Object[existingRow.length + _aggCalls.size()];
+        System.arraycopy(existingRow, 0, row, 0, existingRow.length);
+        for (int i = 0; i < _windowAccumulators.length; i++) {
+          row[i + existingRow.length] = _windowAccumulators[i]._results.get(partitionKey);
+        }
+        rows.add(row);
+      }
+    }
+    _hasReturnedWindowAggregateBlock = true;
+    if (rows.size() == 0) {
```

Comment:
thanks for the suggestion @ankitsultana. Based on the plans I see that the aggregation and the aggregations with group by occur downstream (under) from the window operator. Is this still a concern in that case? Are there specific types of queries you think I should test out to check which behavior makes more sense here (`EOS` vs. empty `TransferableBlock`?

I've added a few tests in my latest commit to check scenarios like empty filter or empty filter with a GROUP BY (validated that these return an empty result). Are these scenarios relevant to your suggestion here?

Reply:
Users can have a query where they first perform windowing in a sub-query, and then later perform aggregation with some filtering. Taking a Uber themed example, say there's a table which has an entry for each Eats order `(order_id, price,  category=[chinese, indian, italian, etc.], food_item=[burger, pizza, etc.])`, a query may be to find how many categories are there such that the second highest order food_item per category has more than 10k orders in the last day.

If there are no such food categories, ideally we should return a single row `[(0)]`, but if we don't return any block above the window operator we may end up returning no rows `[]`. (@walterddr can perhaps confirm this since he had fixed a related issue with empty response in agg operators)

Though this is a very specific example, I think it might be worthwhile to keep an invariant that each Operator would process at least one block.

Reply:
Got it, thanks for explaining the context here. I did look at @walterddr's change to make agg operators return a non-EOS response when rows is empty. This is only done when the group by set is empty (i.e. aggregation without group by query).

```
    if (rows.size() == 0) {
      if (_groupSet.size() == 0) {
        return constructEmptyAggResultBlock();
      } else {
        return TransferableBlockUtils.getEndOfStreamTransferableBlock();
      }
    } else {
      return new TransferableBlock(rows, _resultSchema, DataBlock.Type.ROW);
    }
```

My worry here when it come to returning an empty block is what should I fill in for the values of other columns (e.g. the select columns) for this scenario? Or do you think perhaps we should only return an empty block if there are no other input columns? cc @walterddr for any thoughts on this as well.

e.g. query where I need to worry about what to fill in for `a.col1`:

`SELECT a.col1, SUM(a.col3) OVER(PARTITION BY a.col1) FROM table a;`

Reply:
I think you can pass in an empty container. `TransferableBlock(Collections.emptyList(), _resultSchema, DataBlock.Type.ROW)`.

Reply:
thanks for clarifying, done!

Reply:
i am not sure i follow the discussion here. what exact query are we trying to mitigate?
using postgres, none of these query returns a single row
```
SELECT id, SUM(val) OVER() FROM test where 0 = 1
SELECT id, SUM(val) OVER(PARTITION BY id) FROM test where 0 = 1
SELECT id, SUM(val) OVER(PARTITION BY id ORDER BY val) FROM test where 0 =1
SELECT id, SUM(val) OVER(PARTITION BY id ORDER BY id) FROM test where 0 = 1
```

Reply:
I dug into this a bit more. @ankitsultana please confirm if the scenario I'm going to discuss here matches the scenario you mentioned or not.

So from what I understood, @ankitsultana wanted an empty block to be returned for queries with sub queries with the window function inside the sub query such as:

```
      {
        "sql": "SELECT SUM(count) FROM (SELECT string_col, COUNT(bool_col) OVER() as count FROM {tbl} WHERE string_col = 'a' AND bool_col = false AND int_col > 200)",
        "outputs": [
          [0]
        ]
      },
```

In scenarios where the window inside the sub-query returns 0 rows, the outer query's aggregation should still work and return a result of '0' instead of no rows.

Now I actually tried the above (and will add tests with my latest commit) and found that it doesn't matter if I return an empty row or EOS here. In both scenarios the outer query aggregation returns the correct result.

Based on this @ankitsultana and @walterddr can you confirm if I can change this back to return EOS when the number of rows is 0?

Reply:
I see. In that case either ways works for me. Can we also add this to the documentation of `MultistageOperator::nextBlock`?

Reply:
I'm not sure if adding documentation on this is really necessary. The above works because the outer query uses an intermediate AggregateOperator which does the right thing (i.e. @walterddr 's fix that we discussed above).

Reply:
correct. the tradition of Postgres (as far as I know), is that all aggregate without group should return 1 row. for example
```
SELECT * FROM tbl WHERE val > (SELECT MAX(val) FROM tbl2)
SELECT * FROM tbl WHERE (SELECT count(val) FROM tbl2) > 0
```
both example the SQL relational planner interprets the inner query result as a single value. this can only happen if the agg without group by operators as I mentioned above.

So i think this rule doesn't apply to window, please share more if the analysis above can be in any way improved

Reply:
Yup makes sense! I've changed this back to return an `EOS` since this doesn't apply to window functions and the `AggregateOperator` handles this case.